### ISSUE TITLE
Overfitting

### ISSUE THREAD

**Issue by NikitaKononov** [primepake/wav2lip_288x288#30]
Hello, can you please suggest me any tricks to prevent syncnet overfitting? Decreasing learning rate doesn't help...
![image](https://user-images.githubusercontent.com/69078939/194694012-bc3eff27-aabb-481b-af82-b7566a6ad21e.png)

**Comment by aishoot** [comment]
> Hello, can you please suggest me any tricks to prevent syncnet overfitting? Decreasing learning rate doesn't help... ![image](https://user-images.githubusercontent.com/69078939/194694012-bc3eff27-aabb-481b-af82-b7566a6ad21e.png)

+1

**Comment by aishoot** [comment]
+1

**Comment by NikitaKononov** [comment]
> What dataset are you training on?

Filtered mix of AVS and voxceleb2

**Comment by skyler14** [comment]
Do you have checkpoints at some of these various steps, are the outputs looking intelligible yet? Do you have a way to get in touch, directly we can try a few things interactively and I can probe you about the data and outputs a bit more easily.

**Comment by primepake** [comment]
all problems about overfitting is your training dataset is not general

**Comment by NikitaKononov** [comment]
> all problems about overfitting is your training dataset is not general

No) 
If overfitting phenomenon was so easy to avoid, so many studies and methods would not exist

I think you operate scripts different from this repo, so for your private implementation your message could be right

Overfitting reasons with this repo:
- Syncnet audio encoder presented in this repo is too weak for 288px face encoder, need to add layers. I've done that - it works, but 24/32/48kHz audio input would be ideal
- You need to work properly with your learning rate. With wrong LR you can stuck on saddle points and train for weeks / thousands of epochs
- Dropout layers help to avoid overfitting and make training process more efficient

**Comment by primepake** [comment]
> > all problems about overfitting is your training dataset is not general
> 
> No) If overfitting phenomenon was so easy to avoid, so many studies and methods would not exist
> 
> I think you operate scripts different from this repo, so for your private implementation your message could be right
> 
> Overfitting reasons with this repo:
> 
> * Syncnet audio encoder presented in this repo is too weak for 288px face encoder, need to add layers. I've done that - it works, but 24/32/48kHz audio input would be ideal
> * You need to work properly with your learning rate. With wrong LR you can stuck on saddle points and train for weeks / thousands of epochs
> * Dropout layers help to avoid overfitting and make training process more efficient

Do you mean apply dropout for syncnet or wav2lip?

**Comment by NikitaKononov** [comment]
> > > all problems about overfitting is your training dataset is not general
> > 
> > 
> > No) If overfitting phenomenon was so easy to avoid, so many studies and methods would not exist
> > I think you operate scripts different from this repo, so for your private implementation your message could be right
> > Overfitting reasons with this repo:
> > 
> > * Syncnet audio encoder presented in this repo is too weak for 288px face encoder, need to add layers. I've done that - it works, but 24/32/48kHz audio input would be ideal
> > * You need to work properly with your learning rate. With wrong LR you can stuck on saddle points and train for weeks / thousands of epochs
> > * Dropout layers help to avoid overfitting and make training process more efficient
> 
> Do you mean apply dropout for syncnet or wav2lip?

Syncnet audio encoder. Video encoder performs well even without dropout

**Comment by primepake** [comment]
make senses! That would be perfect with audio_encoder

**Comment by primepake** [comment]
I also trained with dropout layer, the result is pretty good

**Comment by aishoot** [comment]
> I also trained with dropout layer, the result is pretty good

Can you share your dropout layer code?

**Comment by NikitaKononov** [comment]
> I also trained with dropout layer, the result is pretty good

Can you please show you train/eval losses charts?
I think it's not sensitive and private/nda information, I just want to see, how you training behaves
thanks!

**Comment by hannarud** [comment]
@NikitaKononov I can share my training graphs. I'm currently at 0.29 train/eval loss. Not overfitted yet, but nearly there. But still better than 0.39.
What I did - really strictly followed the data preprocessing advice from @primepake here https://github.com/primepake/wav2lip_288x288/issues/21#issuecomment-1172847918.
This is how the beginning of training looked like:
![image](https://user-images.githubusercontent.com/3581666/207385156-5246e70b-77ad-4c98-83c6-304dbad28456.png)
And this is how the last steps looked like:
![image](https://user-images.githubusercontent.com/3581666/207385623-15c244b0-301b-475f-aa74-996ea98a995f.png)

At this point I have two questions:
* @primepake did your training graphs look similar? Can you share them? Was it approaching 0.25 very smooth? I don't see my loss tends to decrease to 0.25, or it will take an enormous amount of time... despite it is still decreasing
* the sync correction you used - was it corrected to exactly 0, or +-3 was acceptable as well? Mine is 0 now and I suspect it's a bit too strict for the model - +-1 frame and `y` gets set to 0 for this case. And as I see it, this is a bit too hard condition. 1 frame apart - and the syncnet should already say "out-of-sync"
* and the third question is connected to the second one - did you change the loss function? There are discussions about that, but I don't understand which loss was used by whom in the end :)

Thank you very much!

**Comment by primepake** [comment]
- Yep! My loss also similar! Also I took 6M to get below at 0.25
- Did you use syncnet_python with correct way? Because Syncnet_python need expand the video from the face area!
- Yes! The changing in loss function is a crucial for this project!

**Comment by primepake** [comment]
> > I also trained with dropout layer, the result is pretty good
> 
> Can you please show you train/eval losses charts? I think it's not sensitive and private/nda information, I just want to see, how you training behaves thanks!

it seems like it dropped sharply in the first steps but after that it stopped so I changed the group norm instead of batch norm

**Comment by hannarud** [comment]
> * Yep! My loss also similar! Also I took 6M to get below at 0.25

Thanks! In this case there's still 2.5 million steps to go :sweat_smile: the loss might decrease to 0.25, because still no clear overfitting and it continues to go down.

> * Did you use syncnet_python with correct way? Because Syncnet_python need expand the video from the face area!

I believe, I used it correctly: applied syncnet to the whole frame, not to the face. Also, I checked the sync-corrected videos again - and if before syncnet_python gave me nonzero offset, after the sync-correction the offset was always 0. I'm a bit concerned by this, actually - maybe, I should have some videos with not so strict sync (I mean, allow +-1 or +-2 frames offset). But currently I applied this sync procedure to the whole dataset and my whole dataset has only zero offset.

> * Yes! The changing in loss function is a crucial for this project!

Yes, I suspect this. Will think about it. Thank you for the response!

**Comment by NikitaKononov** [comment]
> @NikitaKononov I can share my training graphs. I'm currently at 0.29 train/eval loss. Not overfitted yet, but nearly there. But still better than 0.39. What I did - really strictly followed the data preprocessing advice from @primepake here [#21 (comment)](https://github.com/primepake/wav2lip_288x288/issues/21#issuecomment-1172847918). This is how the beginning of training looked like: ![image](https://user-images.githubusercontent.com/3581666/207385156-5246e70b-77ad-4c98-83c6-304dbad28456.png) And this is how the last steps looked like: ![image](https://user-images.githubusercontent.com/3581666/207385623-15c244b0-301b-475f-aa74-996ea98a995f.png)
> 
> At this point I have two questions:
> 
> * @primepake did your training graphs look similar? Can you share them? Was it approaching 0.25 very smooth? I don't see my loss tends to decrease to 0.25, or it will take an enormous amount of time... despite it is still decreasing
> * the sync correction you used - was it corrected to exactly 0, or +-3 was acceptable as well? Mine is 0 now and I suspect it's a bit too strict for the model - +-1 frame and `y` gets set to 0 for this case. And as I see it, this is a bit too hard condition. 1 frame apart - and the syncnet should already say "out-of-sync"
> * and the third question is connected to the second one - did you change the loss function? There are discussions about that, but I don't understand which loss was used by whom in the end :)
> 
> Thank you very much!

Thanks a lot for sharing! Pretty nice graph for ResNet
How many video segments do you have in your training set? 

And did you change any audio parameters in hyperparams?

**Comment by aishoot** [comment]
I also want to ask how many video segments do you have in your training set.

**Comment by NikitaKononov** [comment]
@hannarud 

Thanks a lot for sharing!

Did you modify any hparams? Especially audio
hop_size for example, or smth else

How many video segments do you have in your training set?

Did you set any padding in face detection in preprocessing?
Did you change loss function? Or you use BCELoss?
Did you change activation function to LeakyReLU or PReLU? 

Or you use code in repo as it is (with ReLU and other default parameters)

Looking forward for your answer, thanks

**Comment by wllps1988315** [comment]
> > all problems about overfitting is your training dataset is not general
> 
> No) If overfitting phenomenon was so easy to avoid, so many studies and methods would not exist
> 
> I think you operate scripts different from this repo, so for your private implementation your message could be right
> 
> Overfitting reasons with this repo:
> 
> * Syncnet audio encoder presented in this repo is too weak for 288px face encoder, need to add layers. I've done that - it works, but 24/32/48kHz audio input would be ideal
> * You need to work properly with your learning rate. With wrong LR you can stuck on saddle points and train for weeks / thousands of epochs
> * Dropout layers help to avoid overfitting and make training process more efficient

hi, may I ask you a question?

the Sync loss is below 0.25,
but the wav2lip, the loss is as follows:

Saved checkpoint: checkpoints/checkpoint_step001734000.pth
Evaluating for 700 steps
L1: 0.013941520919356383, Sync loss: 2.2253419873802875
Saved checkpoint: checkpoints/checkpoint_step001737000.pth
Evaluating for 700 steps
L1: 0.014049437306236779, Sync loss: 1.8948244154028138
Starting Epoch: 277
Saved checkpoint: checkpoints/checkpoint_step001740000.pth
Evaluating for 700 steps
L1: 0.014056642805511789, Sync loss: 1.8763052398570084
Saved checkpoint: checkpoints/checkpoint_step001743000.pth
Evaluating for 700 steps
L1: 0.014362473078018929, Sync loss: 1.8959102338969112
Starting Epoch: 278
Saved checkpoint: checkpoints/checkpoint_step001746000.pth
Evaluating for 700 steps
L1: 0.01367571011329131, Sync loss: 1.896067845073134
Saved checkpoint: checkpoints/checkpoint_step001749000.pth
Evaluating for 700 steps
L1: 0.013082690317762987, Sync loss: 1.9725817754418296
Starting Epoch: 279
Saved checkpoint: checkpoints/checkpoint_step001752000.pth
Evaluating for 700 steps
L1: 0.013426973524313426, Sync loss: 2.784595004799703
Saved checkpoint: checkpoints/checkpoint_step001755000.pth
Evaluating for 700 steps
L1: 0.013677941734893262, Sync loss: 1.5971123527698272
Starting Epoch: 280
Saved checkpoint: checkpoints/checkpoint_step001758000.pth
Evaluating for 700 steps
L1: 0.014198551799427086, Sync loss: 1.828956418557786
Saved checkpoint: checkpoints/checkpoint_step001761000.pth
Evaluating for 700 steps
L1: 0.013761360742970132, Sync loss: 2.2420577421763146
Starting Epoch: 281



how can I solved this? change learning rate and how?

**Comment by wllps1988315** [comment]
> > > I also trained with dropout layer, the result is pretty good
> > 
> > 
> > Can you please show you train/eval losses charts? I think it's not sensitive and private/nda information, I just want to see, how you training behaves thanks!
> 
> it seems like it dropped sharply in the first steps but after that it stopped so I changed the group norm instead of batch norm

could you describe how to use syncnet_python with correct way?

**Comment by primepake** [comment]
just input the video
https://github.com/joonson/syncnet_python

**Comment by Erickrus** [comment]
@NikitaKononov did you finally resolve this issue and achieve a lower loss ?

**Comment by piwawa** [comment]
Did you get 0.25 loss?
